File: string/manacher.py
# Manacher's Algorithm
class Manacher():
	def __init__(self, s: str) -> None:
		self.s = s
	def coustruct(self) -> list:
		i, j = 0, 0
		res = [0] * len(self.s)
		while i < len(self.s):
			while i - j >= 0 and i + j < len(self.s) and self.s[i - j] == self.s[i + j]:
				j += 1
			res[i] = j
			k = 1
			while i - k >= 0 and i + k < len(self.s) and k + res[i - k] < j:
				res[i + k] = res[i - k]
				k += 1
			i += k
			j -= k
		return res

File: Cpp/string/manacher.cpp
#include <string>
#include <vector>

std::vector<int> manacher(const std::string& s) {
	int i = 0, j = 0;
	std::vector<int> res((int)s.size(), 0);
	while (i < (int)s.size()) {
		while (i - j >= 0 && i + j < (int)s.size() && s[i - j] == s[i + j]) j++;
		res[i] = j;
		int k = 1;
		while (i - k >= 0 && i + k < (int)s.size() && k + res[i - k] < j) {
			res[i + k] = res[i - k];
			k++;
		}
		i += k;
		j -= k;
	}
	return res;
}